docs/jax-101/02-jitting.ipynb

```diff
@@ -0,0 +1,570 @@
+    "\n",
+    "*Authors: TODO*\n",
+    "\n",
+    "In this section, we will take a further look at how JAX works, and how we can make it performant. In the previous lesson, we discussed that JAX allows us to transform Python functions. This is done by first converting the Python function into a simple intermediate language called jaxpr. The transformations then work on the jaxpr representation. \n",
+    "\n",
+    "We can see the jaxpr of a function by using `jax.make_jaxpr`, as seen in the code below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "id": "P9Xj77Wx3Z2P",
+    "outputId": "77d28eec-bfbb-487b-b10f-e47bc8f7dd0d"
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{ lambda  ; a.\n",
+      "  let b = log a\n",
+      "      c = log 2.0\n",
+      "      d = div b c\n",
+      "  in (d,) }\n"
+     ]
+    }
+   ],
+   "source": [
+    "import jax\n",
+    "import jax.numpy as jnp\n",
+    "\n",
+    "global_list = []\n",
+    "\n",
+    "def log2(x):\n",
+    "  global_list.append(x)\n",
+    "  ln_x = jnp.log(x)\n",
+    "  ln_2 = jnp.log(2.0)\n",
+    "  return ln_x / ln_2\n",
+    "\n",
+    "print(jax.make_jaxpr(log2)(3.0))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "jiDsT7y0RwIp"
+   },
+   "source": [
+    "To understand exactly what the expression above means, check https://jax.readthedocs.io/en/latest/jaxpr.html.\n",
+    "\n",
+    "Importantly, note how the jaxpr does not capture the side-effect of the function: there is nothing in it corresponding to `global_list.append(x)`. This is a feature, not a bug: JAX is designed to understand side-effect-free (a.k.a. functionally pure) code. If *pure function* and *side-effect* are unfamiliar terms, this is explained in a [little more detail](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Pure-functions) in the JAX docs.\n",
+    "\n",
+    "Of course, impure functions can still be written and even run, but JAX gives no guarantees about their behaviour once converted to jaxpr. However, as a rule of thumb, you can expect (but shouldn't rely on) the side-effects of a JAX-transformed function to run once (during the first call), and never again. This is because of the way that JAX generates jaxpr, using a process called 'tracing'.\n",
+    "\n",
+    "When tracing, JAX wraps each argument by a *tracer* object. These tracers then record all JAX operations performed on them during the function call (which happens in regular python). Then, JAX uses the tracer records to reconstruct the entire function. The output of that reconstruction is the jaxpr. Since the tracers do not record the python side-effects, they do not appear in the jaxpr. However, the side-effects still happen during the trace itself.\n",
```

Comment:
Suggestion: nit ("python" 🐍  -> "Python")

```suggestion
    "When tracing, JAX wraps each argument by a *tracer* object. These tracers then record all JAX operations performed on them during the function call (which happens in regular Python). Then, JAX uses the tracer records to reconstruct the entire function. The output of that reconstruction is the jaxpr. Since the tracers do not record the Python side-effects, they do not appear in the jaxpr. However, the side-effects still happen during the trace itself.\n",
```